Lib/typing.py

```diff
@@ -1290,13 +1291,30 @@ def __call__(self, *args, **kwargs):
         if not self._inst:
             raise TypeError(f"Type {self._name} cannot be instantiated; "
                             f"use {self.__origin__.__name__}() instead")
-        result = self.__origin__(*args, **kwargs)
+        clas = self.__origin__
+        is_python_type = inspect.isfunction(clas.__new__)
+
+        if is_python_type:
+            # GH-123777: Some types try and access __orig_class__ in their __init__
+            result = clas.__new__(clas, *args, **kwargs)
+        else:
+            # This is a C type with a custom __new__
+            #
+            # I'm worried that trying to set attributes on C types before they've been
+            # fully initialized will be problematic, so let's just disallow them
+            # from using the __orig_class__ in the initializer for now.
+            result = clas(*args, **kwargs)
+
         try:
             result.__orig_class__ = self
         # Some objects raise TypeError (or something even more exotic)
         # if you try to set attributes on them; we guard against that here
         except Exception:
             pass
+
+        if is_python_type:
```

Comment:
There can be C types with `__init__` as well.

Reply:
As in, there are C types that might want to access their own `__orig_class__`?